homeassistant/components/tuya/cover.py

```diff
@@ -0,0 +1,137 @@
+) -> None:
+    """Set up tuya cover dynamically through tuya discovery."""
+    hass.data[DOMAIN][entry.entry_id][TUYA_HA_TUYA_MAP][
+        DEVICE_DOMAIN
+    ] = TUYA_SUPPORT_TYPE
+
+    async def async_discover_device(dev_ids):
+        """Discover and add a discovered tuya cover."""
+        _LOGGER.debug("cover add-> %s", dev_ids)
+        if not dev_ids:
+            return
+        entities = await hass.async_add_executor_job(
+            _setup_entities, hass, entry, dev_ids
+        )
+        async_add_entities(entities)
+
+    async_dispatcher_connect(
+        hass, TUYA_DISCOVERY_NEW.format(DEVICE_DOMAIN), async_discover_device
+    )
+
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    device_ids = []
+    for (device_id, device) in device_manager.device_map.items():
+        if device.category in TUYA_SUPPORT_TYPE:
+            device_ids.append(device_id)
+    await async_discover_device(device_ids)
+
+
+def _setup_entities(
+    hass: HomeAssistant, entry: ConfigEntry, device_ids: list
+) -> list[Entity]:
+    """Set up Tuya Cover."""
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    entities: list[Entity] = []
+    for device_id in device_ids:
+        device = device_manager.device_map[device_id]
+        if device is None:
+            continue
+        entities.append(TuyaHaCover(device, device_manager))
+        hass.data[DOMAIN][entry.entry_id][TUYA_HA_DEVICES].add(device_id)
+    return entities
+
+
+class TuyaHaCover(TuyaHaEntity, CoverEntity):
+    """Tuya Switch Device."""
+
+    _attr_device_class = DEVICE_CLASS_CURTAIN
+
+    @property
+    def is_closed(self) -> bool | None:
+        """Return is cover is closed."""
+        return None
+
+    @property
+    def current_cover_position(self) -> int:
+        """Return cover current position."""
+        position = self.tuya_device.status.get(DPCODE_PERCENT_STATE, 0)
+        if DPCODE_SITUATION_SET not in self.tuya_device.status:
+            return 1 + int(0.98 * (100 - position))
+        if self.tuya_device.status.get(DPCODE_SITUATION_SET) == "fully_open":
+            return 1 + 0.98 * position
+        
+        return 1 + int(0.98 * (100 - position))
+
+    def open_cover(self, **kwargs: Any) -> None:
+        """Open the cover."""
+        self._send_command([{"code": DPCODE_CONTROL, "value": "open"}])
+
+    def close_cover(self, **kwargs: Any) -> None:
+        """Close cover."""
+        self._send_command([{"code": DPCODE_CONTROL, "value": "close"}])
+
+    def stop_cover(self, **kwargs):
+        """Stop the cover."""
+        self._send_command([{"code": DPCODE_CONTROL, "value": "stop"}])
+
+    def set_cover_position(self, **kwargs):
+        """Move the cover to a specific position."""
+        _LOGGER.debug("cover--> %s", kwargs)
+        self._send_command(
+            [{"code": DPCODE_PERCENT_CONTROL, "value": kwargs[ATTR_POSITION]}]
```

Comment:
Does Tuya and Home Assistant the same position? Above in `current_cover_position` we take `100 - position` to calculate the current position in Home Assistant.

Reply:
Is the same

Reply:
If the Tuya and Home Assistant position representations are the same, why do we take `100 - position` in the `current_cover_position` property to calculate the Home Assistant representation?

Reply:
According to the dp code of the "situation_set" of the device, can judge whether 100% is fully open or fully close. By default, 100% means fully close and 0% means fully open, which can solve the problem of keeping state synchronization between Tuya Smart App and Home App.

Reply:
Ok. So you mean that the device checks DPCODE_SITUATION_SET in the background here when changing position, to know if it should translate position to inverted or not?

Reply:
When Cover is initialized, the DPCODE_SITUATION_SET is used to determine whether the device will use position or 100-position. The rule does not change during use.

Reply:
So then we should be able to use this too to implement `is_closed` property right?